merge.go

```diff
@@ -0,0 +1,474 @@
+	}
+	db.ForEachProperty(func(s string, i interface{}) bool {
+		_, ok := mem.GetProperty(s)
+		if !ok {
+			delta.removeProp = append(delta.removeProp, s)
+		}
+		return true
+	})
+
+}
+
+// compares two lpg nodes and returns a diff
+func compareGraphNode(mem, db *lpg.Node, operationType bool) delta {
+	delta := delta{isNode: true}
+	if db == nil {
+		delta.addLabel = mem.GetLabels().Slice()
+		delta.setProp = ls.CloneProperties(mem)
+		return delta
+	}
+	delta.addLabel = findNodeLabelDiff(mem.GetLabels(), db.GetLabels())
+	// Expected properties must be a subset
+	setPropertiesDelta(delta, mem, operationType)
+	if operationType == overwriteOp {
+		setNodeOverwrite(mem, db, delta)
+	}
+	return delta
+}
+
+// compares two native lpg nodes and returns a diff
+func compareGraphEdge(mem, db *lpg.Edge, operationType bool) delta {
+	delta := delta{isNode: false}
+	if db == nil {
+		delta.addLabel = []string{mem.GetLabel()}
+		delta.setProp = ls.CloneProperties(mem)
+	}
+	if mem.GetLabel() != db.GetLabel() {
+		delta.addLabel = []string{mem.GetLabel()}
+	}
+	setPropertiesDelta(delta, mem, mergeOp)
+	if operationType == overwriteOp {
+		setEdgeOverwrite(mem, db, delta)
+	}
+	return delta
+}
+
+type OperationQueue struct {
+	ops []fmt.Stringer
+}
+
+type updateNode struct {
+	id     int64
+	labels []string
+	props  map[string]interface{}
+}
+
+type createNode struct {
+	id     int64
+	labels []string
+	props  map[string]interface{}
+}
+type createEdge struct {
+	id    int64
+	typ   []string
+	props map[string]interface{}
+}
+
+type updateEdge struct {
+	id    int64
+	typ   []string
+	props map[string]interface{}
+}
+
+func buildQueriesFromOperations(ops []fmt.Stringer) []string {
+	res := make([]string, 0)
+	for _, op := range ops {
+		res = append(res, op.String())
+	}
+	return res
+}
+
+func Merge(memGraph, dbGraph *lpg.Graph, dbGraphIds map[*lpg.Node]int64, dbEdges map[*lpg.Edge]int64) (*lpg.Graph, []string, error) {
+	qs := make([]string, 0)
+	memEntitiesMap := ls.GetEntityInfo(memGraph)
+	dbEntitiesMap := ls.GetEntityInfo(dbGraph)
+	findDBCounterpart := func(n *lpg.Node, schemaPV, idPV *ls.PropertyValue) *lpg.Node {
+		if ls.AsPropertyValue(n.GetProperty(ls.EntitySchemaTerm)).IsEqual(schemaPV) {
+			if ls.AsPropertyValue(n.GetProperty(ls.EntityIDTerm)).IsEqual(idPV) {
+				return n
+			}
+		}
+		return nil
+	}
+	memEntities := make([]*lpg.Node, 0, len(memEntitiesMap))
+	dbEntities := make([]*lpg.Node, 0, len(memEntitiesMap))
+	// align by index matching entity roots i.e, first entity in mem slice should match first entity in db slice
+	for n := range memEntitiesMap {
+		schemaPV := ls.AsPropertyValue(n.GetProperty(ls.EntitySchemaTerm))
+		eidPV := ls.AsPropertyValue(n.GetProperty(ls.EntityIDTerm))
+		memEntities = append(memEntities, n)
+		for db := range dbEntitiesMap {
+			if findDBCounterpart(db, schemaPV, eidPV) != nil {
+				dbEntities = append(dbEntities, db)
+				delete(dbEntitiesMap, db)
+			}
+		}
+	}
```

Comment:
If no match, add nil to the list. Better: initialize the slices with fixed size, and set by index